// casper-storage/dist/tsc/browser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
/** ! Entrypoint of browser builds */
__exportStar(require("./index"), exports);
const index_1 = require("./index");
const index_2 = require("./index");
const index_3 = require("./index");
const index_4 = require("./index");
const index_5 = require("./index");
// Map public classes to browser global object
if (typeof window !== 'undefined') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const register = window;
    const CasperStorage = {
        // Utilities
        TypeUtils: index_1.TypeUtils,
        // Key - mnemophic
        KeyFactory: index_2.KeyFactory,
        // crypto
        EncryptionType: index_3.EncryptionType,
        CryptoUtils: index_3.CryptoUtils,
        AESUtils: index_3.AESUtils,
        EncoderUtils: index_3.EncoderUtils,
        KeyParser: index_2.KeyParser,
        // Wallets
        CasperWallet: index_4.CasperWallet,
        CasperHDWallet: index_4.CasperHDWallet,
        CasperLegacyWallet: index_4.CasperLegacyWallet,
        User: index_5.User,
        StorageManager: index_5.StorageManager
    };
    register.CasperStorage = CasperStorage;
}
